pyccel/parser/semantic.py

```diff
@@ -964,6 +964,19 @@ def _handle_function(self, expr, func, args, **settings):
                 errors.report("Too few arguments passed in function call",
                         symbol = expr,
                         severity='error')
+            elif isinstance(func, FunctionDef):
+                matching_types = all(
+                         input_arg.dtype     is func_arg.dtype and \
+                         input_arg.precision == func_arg.precision and \
+                        (input_arg.rank      == func_arg.rank     or func.is_elemental) \
+                        for input_arg, func_arg in zip(new_expr.args, func.arguments) \
+                            if input_arg is not Nil() and \
+                                not isinstance(func_arg, FunctionAddress) and \
+                                func_arg.dtype is not NativeGeneric())
+                if not matching_types:
+                    errors.report(INCOMPATIBLE_ARGUMENT,
+                            symbol = expr,
+                            severity='error')
```

Comment:
Can't this be in a small function or precise witch argument is incompatible?

Reply:
I agree that a function makes sense. Do you think it should be a member of `FunctionDef` or `SemanticParser`?

> or precise witch argument is incompatible?

The reason I did not do this was because I was worried that the code might be slow as this is checked at every function call. But I did not benchmark it. As long as there are not too many arguments/the function call does not appear too often, it should not really cause problems. I will improve this :+1:

Reply:
did you try to use __handle_function_args_ loop?

Reply:
The function `_handle_function_args` has no knowledge of the `FunctionDef` and it is also used for other objects such as `PyccelInternalFunction`. If I checked it in the for loop then I would need to add an `if isinstance(func, FunctionDef)` inside the for loop. This is something that should be avoided at all cost from a performance point of view

Reply:
Fixed in a799243